### DISCUSSION TITLE
Flux controller ecosystem extensibility

### DISCUSSION THREAD

**Discussion by scottrigby** [fluxcd/flux2#621] [General]
Flux has introduced a toolkit, on which the current Flux-maintained controllers are build, which will also allow the wider community to build additional controllers for Flux.

Can we make it easier for the community to build, share and discover these controllers? What criteria if any should there be for promoting them?

We discussed this in today's Flux dev meeting ([see notes](https://docs.google.com/document/d/1l_M0om0qUEN_NNiGgpqJ2tvsF2iioHkaARDeh6b70B0/edit#heading=h.pv5sgx1t62q8)), but opening this general discussion for follow-up and additional community feedback.

**Comment by kingdonb** [comment]
This question about what to do when controllers are contributed or new controllers are contributed that might be beneficial to the community remains relevant

How do we promote new controllers when they come up? What kind of ideas are people thinking about implementing as controllers, and can they have a place in the Flux community?

A great example is: https://github.com/chanwit/tf-controller from @chanwit which is new from very recently, I think I heard this will get a place in the https://github.com/fluxcd-community organization 🎉 ✨ 💯 🥇 

Can we support those who are building such controllers on the GitOps Toolkit even further with tools like [controller-check](https://github.com/darkowlzz/controller-check) (@darkowlzz) to ensure conformance with cluster health requirements or even more specific Flux requirements? What else is required in order to make sure that your controllers that integrate with the toolkit conform to the design and how should we provide guidance to those controllers to test that they remain compatible with future releases of Flux?

There's also the issue of promotion - short of migrating your repository to the FluxCD orgs, what else can repo owners do to tag your repo #gitops-toolkit and make it more discoverable by the Flux community? There's this GitHub Topics feature which we were a bit skeptical about when this came up at Bug Scrub just now, but actually looking at it this looks like a pretty helpful entrypoint now:

* https://github.com/topics/gitops
* https://github.com/topics/gitops-toolkit
